### ISSUE TITLE
[@expo/fingerprint] Non-deterministic output (relative config plugins + expo-modules-autolinking)

### ISSUE THREAD

**Issue by ItsWendell** [expo/expo#24069]
### Minimal reproducible example

https://github.com/ItsWendell/expo-fingerprint-non-deterministic-reproducible

### Summary
Github Issue: https://github.com/expo/expo/issues/24069

# Non-deterministic `@expo/fingerprint`

The package `@expo/fingerprint` is supposed to create a deterministic fingerprint of an Expo project. I was attempting to use this to fingerprint the runtime version, but ran into two issues.

To run fingerprinting in this repository:

1. Install packages: `pnpm install`
2. Open your terminal in the `apps/expo` folder
3. Run `pnpm fingerprint`

This starts a script that runs the fingerprint package, writes the output to the fingerprint.json file, and outputs the diff between the previous version.

## Issue 1: Relative config plugin paths in app config
Having an relative plugin path results in the following diff always being non-deterministic:

```json
[
  {
    "type": "dir",
    "filePath": "",
    "reasons": [
      "expoConfigPlugins",
      "expoConfig",
      "easBuild"
    ],
    "hash": "15a1ebf6d71be03c5b682712cc5864fe8ab54a73"
  }
]
```

To reproduce this, in `apps/expo/app.config.ts` add / remove `plugins: ["expo-router", "./expo-plugins/with-modify-gradle.js"]`, and run `pnpm fingerprint`` in this folder multiple times.

## Issue 2: androidAutoLinking / expo-modules-autolinking

Now, if we remove this relative plugin path, and run `pnpm fingerprint` multiple times, it results in the following issue with `expoAutolinkingAndroid`, here's the output I get:

1. A couple (4/5) non-diff `[]` outputs
2. The following diff:
```json
[
  {
    "type": "contents",
    "id": "expoAutolinkingConfig:android",
    "contents": "{\"extraDependencies\":{\"androidMavenRepos\":[],\"iosPods\":{}},\"modules\":[{\"packageName\":\"expo\",\"packageVersion\":\"49.0.7\",\"projects\":[{\"name\":\"expo\",\"sourceDir\":\"../../node_modules/expo/android\"}],\"modules\":[]},{\"packageName\":\"expo-application\",\"packageVersion\":\"5.3.0\",\"projects\":[{\"name\":\"expo-application\",\"sourceDir\":\"../../node_modules/expo-application/android\"}],\"modules\":[]},{\"packageName\":\"expo-constants\",\"packageVersion\":\"14.4.2\",\"projects\":[{\"name\":\"expo-constants\",\"sourceDir\":\"../../node_modules/expo-constants/android\"}],\"modules\":[\"expo.modules.constants.ConstantsModule\"]},{\"packageName\":\"expo-file-system\",\"packageVersion\":\"15.4.3\",\"projects\":[{\"name\":\"expo-file-system\",\"sourceDir\":\"../../node_modules/expo-file-system/android\"}],\"modules\":[\"expo.modules.filesystem.FileSystemModule\"]},{\"packageName\":\"expo-font\",\"packageVersion\":\"11.4.0\",\"projects\":[{\"name\":\"expo-font\",\"sourceDir\":\"../../node_modules/expo-font/android\"}],\"modules\":[]},{\"packageName\":\"expo-keep-awake\",\"packageVersion\":\"12.3.0\",\"projects\":[{\"name\":\"expo-keep-awake\",\"sourceDir\":\"../../node_modules/expo-keep-awake/android\"}],\"modules\":[]},{\"packageName\":\"expo-modules-core\",\"packageVersion\":\"1.5.9\",\"projects\":[{\"name\":\"expo-modules-core\",\"sourceDir\":\"../../node_modules/expo-modules-core/android\"},{\"name\":\"expo-modules-core$android-annotation\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation\"},{\"name\":\"expo-modules-core$android-annotation-processor\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation-processor\"}],\"modules\":[]},{\"packageName\":\"expo-splash-screen\",\"packageVersion\":\"0.20.5\",\"projects\":[{\"name\":\"expo-splash-screen\",\"sourceDir\":\"../../node_modules/expo-splash-screen/android\"}],\"modules\":[\"expo.modules.splashscreen.SplashScreenModule\"]}]}",
    "reasons": [
      "expoAutolinkingAndroid"
    ],
    "hash": "1f146ce0760f451f65844001ad5ad6f8e0da847f"
  }
]
```
3. A couple of non-diff outputs (4/5) `[]`
4. Another diff output with the following
```json
[
  {
    "type": "dir",
    "filePath": "../../node_modules/expo-modules-core/android",
    "reasons": [
      "expoAutolinkingAndroid"
    ],
    "hash": "2bdad231cf587685afe1882ea2ecf965f4c2df4c"
  },
  {
    "type": "contents",
    "id": "expoAutolinkingConfig:android",
    "contents": "{\"extraDependencies\":{\"androidMavenRepos\":[],\"iosPods\":{}},\"modules\":[{\"packageName\":\"expo\",\"packageVersion\":\"49.0.7\",\"projects\":[{\"name\":\"expo\",\"sourceDir\":\"../../node_modules/expo/android\"}],\"modules\":[]},{\"packageName\":\"expo-application\",\"packageVersion\":\"5.3.0\",\"projects\":[{\"name\":\"expo-application\",\"sourceDir\":\"../../node_modules/expo-application/android\"}],\"modules\":[]},{\"packageName\":\"expo-constants\",\"packageVersion\":\"14.4.2\",\"projects\":[{\"name\":\"expo-constants\",\"sourceDir\":\"../../node_modules/expo-constants/android\"}],\"modules\":[\"expo.modules.constants.ConstantsModule\"]},{\"packageName\":\"expo-file-system\",\"packageVersion\":\"15.4.3\",\"projects\":[{\"name\":\"expo-file-system\",\"sourceDir\":\"../../node_modules/expo-file-system/android\"}],\"modules\":[\"expo.modules.filesystem.FileSystemModule\"]},{\"packageName\":\"expo-font\",\"packageVersion\":\"11.4.0\",\"projects\":[{\"name\":\"expo-font\",\"sourceDir\":\"../../node_modules/expo-font/android\"}],\"modules\":[]},{\"packageName\":\"expo-keep-awake\",\"packageVersion\":\"12.3.0\",\"projects\":[{\"name\":\"expo-keep-awake\",\"sourceDir\":\"../../node_modules/expo-keep-awake/android\"}],\"modules\":[]},{\"packageName\":\"expo-modules-core\",\"packageVersion\":\"1.5.9\",\"projects\":[{\"name\":\"expo-modules-core$android-annotation\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation\"},{\"name\":\"expo-modules-core\",\"sourceDir\":\"../../node_modules/expo-modules-core/android\"},{\"name\":\"expo-modules-core$android-annotation-processor\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation-processor\"}],\"modules\":[]},{\"packageName\":\"expo-splash-screen\",\"packageVersion\":\"0.20.5\",\"projects\":[{\"name\":\"expo-splash-screen\",\"sourceDir\":\"../../node_modules/expo-splash-screen/android\"}],\"modules\":[\"expo.modules.splashscreen.SplashScreenModule\"]}]}",
    "reasons": [
      "expoAutolinkingAndroid"
    ],
    "hash": "8cdb71db7052474a3a600a80cfdd9645536e3338"
  }
]
```
5. A couple of non-diff outputs `[]`
6. Another 'failed' run with the following diff:
```json
[
  {
    "type": "dir",
    "filePath": "../../node_modules/expo-modules-core/android-annotation",
    "reasons": [
      "expoAutolinkingAndroid"
    ],
    "hash": "26f281ea4437881518ab1765ee155868eb0fd925"
  },
  {
    "type": "contents",
    "id": "expoAutolinkingConfig:android",
    "contents": "{\"extraDependencies\":{\"androidMavenRepos\":[],\"iosPods\":{}},\"modules\":[{\"packageName\":\"expo\",\"packageVersion\":\"49.0.7\",\"projects\":[{\"name\":\"expo\",\"sourceDir\":\"../../node_modules/expo/android\"}],\"modules\":[]},{\"packageName\":\"expo-application\",\"packageVersion\":\"5.3.0\",\"projects\":[{\"name\":\"expo-application\",\"sourceDir\":\"../../node_modules/expo-application/android\"}],\"modules\":[]},{\"packageName\":\"expo-constants\",\"packageVersion\":\"14.4.2\",\"projects\":[{\"name\":\"expo-constants\",\"sourceDir\":\"../../node_modules/expo-constants/android\"}],\"modules\":[\"expo.modules.constants.ConstantsModule\"]},{\"packageName\":\"expo-file-system\",\"packageVersion\":\"15.4.3\",\"projects\":[{\"name\":\"expo-file-system\",\"sourceDir\":\"../../node_modules/expo-file-system/android\"}],\"modules\":[\"expo.modules.filesystem.FileSystemModule\"]},{\"packageName\":\"expo-font\",\"packageVersion\":\"11.4.0\",\"projects\":[{\"name\":\"expo-font\",\"sourceDir\":\"../../node_modules/expo-font/android\"}],\"modules\":[]},{\"packageName\":\"expo-keep-awake\",\"packageVersion\":\"12.3.0\",\"projects\":[{\"name\":\"expo-keep-awake\",\"sourceDir\":\"../../node_modules/expo-keep-awake/android\"}],\"modules\":[]},{\"packageName\":\"expo-modules-core\",\"packageVersion\":\"1.5.9\",\"projects\":[{\"name\":\"expo-modules-core\",\"sourceDir\":\"../../node_modules/expo-modules-core/android\"},{\"name\":\"expo-modules-core$android-annotation\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation\"},{\"name\":\"expo-modules-core$android-annotation-processor\",\"sourceDir\":\"../../node_modules/expo-modules-core/android-annotation-processor\"}],\"modules\":[]},{\"packageName\":\"expo-splash-screen\",\"packageVersion\":\"0.20.5\",\"projects\":[{\"name\":\"expo-splash-screen\",\"sourceDir\":\"../../node_modules/expo-splash-screen/android\"}],\"modules\":[\"expo.modules.splashscreen.SplashScreenModule\"]}]}",
    "reasons": [
      "expoAutolinkingAndroid"
    ],
    "hash": "1f146ce0760f451f65844001ad5ad6f8e0da847f"
  }
]
```

Diving deeper, it seems that the CLI tool for `expo-modules-autolinking` might be causing this, since this is where it is fetching this information.

### Environment

```
    System:
      OS: Linux 6.4 Fedora Linux 38 (Workstation Edition)
      Shell: 5.9 - /usr/bin/zsh
    Binaries:
      Node: 18.17.1 - /usr/bin/node
      npm: 9.6.7 - /usr/bin/npm
    SDKs:
      Android SDK:
        API Levels: 30, 31, 32, 33
        Build Tools: 30.0.2, 30.0.3, 31.0.0, 32.0.0, 33.0.0, 33.0.2, 34.0.0
        System Images: android-21 | Intel x86 Atom, android-21 | Google APIs Intel x86 Atom, android-24 | Google APIs Intel x86 Atom, android-24 | Google Play Intel x86 Atom, android-27 | Google APIs Intel x86 Atom, android-27 | Google Play Intel x86 Atom, android-28 | Google APIs Intel x86 Atom, android-28 | Google APIs Intel x86 Atom_64, android-28 | Google Play Intel x86 Atom_64, android-29 | Android TV Intel x86 Atom, android-29 | Google Play Intel x86 Atom, android-29 | Google Play Intel x86 Atom_64, android-30 | Google APIs Intel x86 Atom_64, android-30 | Google Play Intel x86 Atom, android-30 | Google Play Intel x86 Atom_64, android-30 | Google APIs ATD Intel x86 Atom, android-31 | Google APIs Intel x86_64 Atom, android-31 | Google Play Intel x86 Atom_64, android-32 | Google APIs Intel x86_64 Atom, android-32 | Google Play Intel x86 Atom_64, android-33 | Google APIs Intel x86_64 Atom, android-33 | Google Play Intel x86 Atom_64
    Expo Workflow: managed
```

**Comment by Kudo** [comment]
hi there! thanks for the great repro.
for issue 2, it should be resolved by #23737. sorry that i didn't publish a new version after the pr. now `@expo/fingerprint@0.0.10` should help. i will take a look for issue 1 later this week.

**Comment by ItsWendell** [comment]
@Kudo what could be reasons for a fingerprint to be different on different machines? I'm the environment mentioned above, vs (the latest github runner for ubuntu)[https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md], I'm getting different hashes.

**Comment by Kudo** [comment]
> issue 1: non-deterministic output when enabling relative config plugin

by design, when there are relative config plugins, we will source the whole app directory. since you generate `fingerprint.json` in the app directory and update its content. it will generate difference hash. you could either move the `fingerprint.json` out from the app directory, or use the `dirExcludes` to exclude some directories:

```patch
--- a/apps/expo/fingerprint.mjs
+++ b/apps/expo/fingerprint.mjs
@@ -5,15 +5,20 @@ import * as path from 'node:path';
 const projectRoot = process.cwd();

 const run = async () => {
-    const previousFingerprint = await fs.readFile(path.join(projectRoot, 'fingerprint.json'), 'utf8').then(JSON.parse).catch(() => null);
+    const fingerprintDir = 'fingerprint';
+    const fingerprintResultPath = path.join(projectRoot, fingerprintDir, 'fingerprint.json');
+    try {
+      await fs.mkdir(path.join(projectRoot, fingerprintDir));
+    } catch {}
+    const previousFingerprint = await fs.readFile(fingerprintResultPath, 'utf8').then(JSON.parse).catch(() => null);

-    const newFingerprint = await createFingerprintAsync(projectRoot);
+    const newFingerprint = await createFingerprintAsync(projectRoot, { dirExcludes: [fingerprintDir] });

     // Write to JSON file
-    await fs.writeFile(path.join(projectRoot, 'fingerprint.json'), JSON.stringify(newFingerprint, null, 2));
+    await fs.writeFile(fingerprintResultPath, JSON.stringify(newFingerprint, null, 2));

     if (previousFingerprint) {
-        const diff = await diffFingerprintChangesAsync(previousFingerprint, projectRoot);
+        const diff = await diffFingerprintChangesAsync(previousFingerprint, projectRoot, { dirExcludes: [fingerprintDir] });

         console.log(JSON.stringify(diff, null, 2));
     } else {
```

> different fingerprint on different machines

good catch for this. this is an issue coming from monorepo and the fingerprint contains unhandled absolute paths, for example [this one](https://github.com/ItsWendell/expo-fingerprint-non-deterministic-reproducible/blob/72fe8633c691ef5145ec3a72cf7a950a6762c1ac/apps/expo/fingerprint.json#L238).

i will address the monorepo issue.

also may be good to have this function and not to calculate fingerprint twice for your case
```ts
async function diffFingerprintChangesAsync(
  fingerprint1: Fingerprint,
  fingerprint2: Fingerprint
): Promise<FingerprintSource[]>;
```

**Comment by Kudo** [comment]
published `@expo/fingerprint@0.1.0` with the fix.
also adding the `diffFingerprints` function so you could just diff two fingerprints without re-calculate the fingerprint again.

```ts
const previousFingerprint = await fs.readFile(fingerprintResultPath, 'utf8').then(JSON.parse).catch(() => null);
const newFingerprint = await createFingerprintAsync(projectRoot, { dirExcludes: [fingerprintDir] });
if (previousFingerprint) {
  const diff = diffFingerprints(previousFingerprint, newFingerprint);
  console.log(JSON.stringify(diff, null, 2));
}
```
